command/meta_backend.go

```diff
@@ -562,6 +562,76 @@ func (m *Meta) backendFromConfig(opts *BackendOpts) (backend.Backend, tfdiags.Di
 	}
 }
 
+// backendFromState returns the initialized (not configured) backend directly
+// from the state. This should be used only when a user runs `terraform init
+// -backend=false`. This function returns a local backend if there is no state
+// or no backend configured.
+func (m *Meta) backendFromState() (backend.Backend, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	// Get the path to where we store a local cache of backend configuration
+	// if we're using a remote backend. This may not yet exist which means
+	// we haven't used a non-local backend before. That is okay.
+	statePath := filepath.Join(m.DataDir(), DefaultStateFilename)
+	sMgr := &state.LocalState{Path: statePath}
+	if err := sMgr.RefreshState(); err != nil {
+		diags = diags.Append(fmt.Errorf("Failed to load state: %s", err))
+		return nil, diags
+	}
+	s := sMgr.State()
+	if s == nil {
+		// no state, so return a local backend
+		log.Printf("[TRACE] Meta.Backend: backend has not previously been initialized in this working directory")
+		s = terraform.NewState()
+		return backendLocal.New(), diags
+	} else if s.Backend != nil {
+		log.Printf("[TRACE] Meta.Backend: working directory was previously initialized for %q backend", s.Backend.Type)
+	} else if s.Backend == nil {
```

Comment:
Somewhat nit, but if you reversed the order of these last two `if`s, do you get to drop one? ex. `if x == nil {}` then the log line doesn't have to be wrapped in an `if`?

Reply:
Not quite, not without more rewriting: there are 3 `if`s there so the 'else' needs to be specific or it will relate to the first `if`. 

It's code I copied from elsewhere and could have spent more time polishing it, so let me go make that clearer, thank you!

Reply:
Your next change is what I was thinking of! Nice fixup